lib/streamlit/connections/sql_connection.py

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from contextlib import contextmanager
+from datetime import timedelta
+from typing import Iterator, Optional, Union, cast
+
+import pandas as pd
+import sqlalchemy
+from sqlalchemy.engine.base import Engine
+from sqlalchemy.orm import Session
+
+from streamlit.connections import BaseConnection
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+REQUIRED_CONNECTION_PARAMS = {"dialect", "username", "host", "port"}
+
+
+class SQL(BaseConnection[Engine]):
+    _default_connection_name = "sql"
+
+    def connect(self, autocommit: bool = False, **kwargs) -> Engine:
+        self._closed = False
+
+        secrets = self.get_secrets()
+
+        if "url" in secrets:
+            url = secrets["url"]
+        else:
+            for p in REQUIRED_CONNECTION_PARAMS:
+                if p not in secrets:
+                    raise StreamlitAPIException(f"Missing SQL DB connection param: {p}")
+
+            driver = f"+{secrets['driver']}" if "driver" in secrets else ""
+            password = f":{secrets['password']}" if "password" in secrets else ""
+            database = f"/{secrets['database']}" if "database" in secrets else ""
+
+            url = "".join(
+                [
+                    secrets["dialect"],
+                    driver,
+                    "://",
+                    secrets["username"],
+                    password,
+                    "@",
+                    secrets["host"],
+                    ":",
+                    secrets["port"],
+                    database,
+                ]
+            )
```

Comment:
Nit: I would vote we use `sqlalchemy.engine.URL.create()` for this if we can instead of manual concat?

Reply:
Ah, yeah, that'd be much nicer 😅

Reply:
done!